### DISCUSSION TITLE
How to handle checkbox?

### DISCUSSION THREAD

**Discussion by jdgamble555** [open-circle/valibot#254] [Q&A] [answered]
I have a checkbox `completed` that is passed to the server.  The value seems to be: `'on' | null`;

How can I handle this with valibot?

I need to validate a boolean value.

```ts
const TaskSchema = object({
    name: string(),
    completed: boolean()
});

const result = safeParse(TaskSchema, Object.fromEntries(formData));
  
...
```

J

**Comment by fabian-hiller** [comment, marked as answer]
I recommend to use `decode-formdata` for that: https://github.com/fabian-hiller/decode-formdata

**Reply by jdgamble555** [reply]
If I don't want to import a whole package to add extra overhead, is there any way to do something like this?

https://github.com/airjp73/remix-validated-form/blob/f430df7c2ee743bc2017831892afca5472f30f35/packages/zod-form-data/src/helpers.ts#L79-L83


J

**Reply by fabian-hiller** [reply]
You have several options. With `coerce` you can change the incoming data before validation to force a boolean. `transform` does the same, but after validation. `union`, as in Zod, is also available for Valibot.

```ts
const Schema1 = coerce(boolean(), Boolean);
const Schema2 = transform(string(), Boolean);
```

**Reply by jdgamble555** [reply]
Thanks, but I'm not understanding this implementation? 

![Screenshot 2023-11-19 092614](https://github.com/fabian-hiller/valibot/assets/8333090/67ea4220-ad0a-41c3-b1a4-8acf4852e53a)

**Reply by fabian-hiller** [reply]
In your case I recommend to use `coerce`. Please give me more details so that I can answer you more appropriately.

**Reply by jdgamble555** [reply]
What more details do you need? It is just a checkbox passed to the server (from above) with the code:
```ts
const result = safeParse(TaskSchema, Object.fromEntries(formData));
```
I'm not understanding the usage of either (the docs don't help much there).

J

**Reply by fabian-hiller** [reply]
Please ask specific questions if this information is not sufficient for you.

- https://valibot.dev/guides/methods/#coerce
- https://valibot.dev/guides/methods/#transform

**Reply by jdgamble555** [reply]
Hi Fabian,

I apologize, but I thought I did. I gave you my entire code. It turns out, I just needed to post the coerce instead of the transform like you put:
```ts
const TodoSchema = object({
    name: string([minLength(1, 'Name is required')]),
    completed: coerce(boolean(), Boolean)
});
```
This worked.

Thanks,

J

**Comment by CanRau** [comment]
another option is rewriting zfd's checkbox
```ts
type CheckboxOpts = {
  trueValue?: string;
};

export const checkbox = ({ trueValue = "on" }: CheckboxOpts = {}) =>
  v.union([
    v.pipe(v.literal(trueValue), v.transform(() => true)),
    v.pipe(v.undefined(), v.transform(() => false))
  ]);
```

or you could allow `boolean`s as well
```ts
export const checkbox = ({ trueValue = "on" }: CheckboxOpts = {}) =>
  v.union([
    v.boolean(),
    v.pipe(v.literal(trueValue), v.transform(() => true)),
    v.pipe(v.undefined(), v.transform(() => false))
  ]);
```

here's a [playground](https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6YBPMAUzgGEALNgYwDWVAB4B5MDHQBeOAG96cODCgBXNgDUyagPwAuOKmXAAdgHMGAXwb02wyLDh8Ixw496CRcGQApZS1RpaHDIARM4hcBb6PPxCEGIS0nIWAJReAHwKmAB0KsbAzt4A2lmKGNlgwOze5eQwbFBk3spqmqRqKQA0OcrILtjQIN7eaVLp-h1dpTmV1eV5ACZs2CZsCyPd5b39g8Oj49hkqGwpKVkAuinWtvbwTi537nHCAExecL4Tge3BcGHGESiXCeInEkneslSGSy83yhRKijK2SoEFIbD6G2m5VmbBq2TqDSaLW+HU22W2qAGUCGIwyX1OnSxFSquNhSxWxjWGx6jR21L2dMOpGOpwuV0Y91cAGU+LwQMh3uUIBQAFb8GC+LJOcBo+oLfSy2IiF6Y1LWSWotj4iCmPGoZDYNgABWQUGO3hlcuQ3T82rAurW+kIzkIkVO5ucqEt1tt5XtjpdbtxnrY8p9jnw-rYev0i2WqwWYfFFrRMbtDudrvdKbTcgzOuzgZwRw4qWLkejpBt5YTVeThtrvszAf1XyLtCAA)

or even more general [`boolish`](https://valibot.dev/playground/?code=JYWwDg9gTgLgBAKjgQwM5wG5wGZQiOAcg2QBtgAjCGQgbgCh6YBPMAUzgCEIJzUALAPJgY6ALxwA3vThwYUAK4x+zAGpkFbVAH4AXHACCUKMmYAeVPOAA7AOYA+BgF8G9NgA9IsOAGMI1yzgqXmABOAkACkk5RWU1DS1wuABtACJ-VIBdOCd9bhCBYVEkyScASnD7GUwAOjBgdgiMGoVrAGtrCAB3awiygBpa+WQA7GgQCIibMCUKsXspatkoNhgFKGsuHlI2EanrGfgAMiO4fcPwiXlNOAAfW5ilFXVSTVQa1Hw2CPh5uUuJNNZmUytVymVXH4AvAAMo+fhsEDIJLNCAUABWbB8MCi1T84B2MDYABN9ME+Pwoo84i83vo0hlBqlmFosuV+vRyq4WOw4BAlBcJM0AJLWbBsKCCAVKMw8tgQbBwOEIpFVehQz47GqkCC2JofZDigAKyCgqG+ysRyEG0XxYEJJP0hH8hByZQ+Ch8Pi0qAh6v8mrY2t1+tQhrYJrNFvhVptvnw9tWjqILNQrvKHq9Pr9Gt4QZ1euaYeNpvNEUtSLjdodpLgrWJbGwNhJbsz3tQvoYua1BdD4cjZYr1qk8YJSdr2DI5tbqE97c7-oCeeDhYNJaj5ZjlZH1fH+muHAzs6zHYhQA) ^^
